### ISSUE TITLE
Training code?

### ISSUE THREAD

**Issue by kabachuha** [bytedance/1d-tokenizer#1]
Hi!

It's an extremely nice work. Do you have training code release in plans?

**Comment by cornettoyu** [comment]
Hi,

Thanks for your interests. For training code and beyond, we are still working on the internal approval process. Before that, feel free to let me know if you have any questions regarding technical details and I am more than happy to address them :)

**Comment by reedscot** [comment]
Awesome work! May I ask which VQGAN implementation was used for the proxy codes?

**Comment by cornettoyu** [comment]
> Awesome work! May I ask which VQGAN implementation was used for the proxy codes?

Thanks for your interests! For proxy codes at warm-up training, we used MaskGIT-VQGAN, the original implementation was in Jax and can be found at https://github.com/google-research/maskgit  We used the pytorch version from huggingface's [open-muse](https://github.com/huggingface/open-muse) which provides a pytorch reimplementation and weight ported from Jax.

**Comment by vkramanuj** [comment]
I have a detailed question about the proxy codes. The Maskgit VQGAN provides a fixed length set of codes (256 or 1024). How do you distill that into 32 or 64 codes during the warm-up procedure for the smaller models? Perhaps I'm misunderstanding the paper. Thanks!

**Comment by cornettoyu** [comment]
> Thanks for the great work!
> 
> I have a detailed question about the proxy codes. The Maskgit VQGAN provides a fixed length set of codes (256 or 1024). How do you distill that into 32 or 64 codes during the warm-up procedure for the smaller models? Perhaps I'm misunderstanding the paper. Thanks!

The MaskGIT-VQGAN's code is used to supervise the output of TiTok's de-tokenizer, similar to BEIT. As we use the mask token sequence (BERT, MAE style) to reconstruct the target sequence, it does not matter how many tokens we are using or they are using. We do not apply any "distill" or "loss" between TiTok's codebook/embedding and MaskGIT-VQ's codebook/embedding etc. Hope it addresses your question.

**Comment by jeasinema** [comment]
Thank you for the reply. I think I get it, but please correct me if I am wrong here -- so during the warm-up stage, eq. (4) in the main paper is different: 1) it produces the codes of MaskGIT-VQGAN instead of pixels directly; 2) the codes will be processed by MaskGIT-VQGAN decoder into pixels.

I would appreciate it if you could update the main paper with an equation clearing these differences if this is the case. That will help the readers a lot!

**Comment by JosephPai** [comment]
@jeasinema 
Based on my understanding of the code, I think you are correct.

In the `decode()` function of the `modeling/titok.py` file, `z_quantized` is first decoded into `decoded_latent`, then being processed to `quantized_states`. After that, the `pixel_decoder` decodes the `quantized_states` into pixel space.

`quantized_states` is the proxy code in the space of VQGAN.
`self.decoder` is the de-tokenizer trained in this paper.
`self.pixel_decoder` is adapted from MaskGiT-VQGAN decoder.

@cornettoyu Please correct if I'm wrong. Thanks.

**Comment by chmxu** [comment]
> @jeasinema Based on my understanding of the code, I think you are correct.
> 
> In the `decode()` function of the `modeling/titok.py` file, `z_quantized` is first decoded into `decoded_latent`, then being processed to `quantized_states`. After that, the `pixel_decoder` decodes the `quantized_states` into pixel space.
> 
> `quantized_states` is the proxy code in the space of VQGAN. `self.decoder` is the de-tokenizer trained in this paper. `self.pixel_decoder` is adapted from MaskGiT-VQGAN decoder.
> 
> @cornettoyu Please correct if I'm wrong. Thanks.

Hi @jeasinema @JosephPai @reedscot , I wonder if you have successfully reproduced the training of the tokenizer?  As I can understand, I need to regress the predicted ```quantized_states``` with the ground truth produced by pretrained MaskGiT. Is that right? If so, what happens when I want to train this model on some specific domain on which MaskGiT may not perform well?

**Comment by cornettoyu** [comment]
Hi,

Thank you all for your interests in our work and patience. We have released the two-stage training code and feel free to check it.
